### PR TITLE
Computing hash of step and materializer works in notebooks

### PR DESCRIPTION
**Pull request by htahir1**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
This is a draft now now. 

# Problem
`inspect.getsource` does not work in jupyter notebooks because:  ipython or Jupyter notebook, the classes/functions defined or the __main__ module does not have a __file__ attribute associated with them so inspect wasn't able to retrieve a source file.

# What I did
Basically I monkeypatched the `inspect.getfile` method following this [SO thread](https://stackoverflow.com/questions/51566497/). However I tried to reverse the monkeypatch in the function itself to retain `inspect.getfile` original integrity. 

FYI `inspect.getfile` is used in `inspect.getsource` and is the breaking call.

# I am looking for input of the following things:

* How to write tests for such a functionality? How can I simulate being in a notebook setting? I have verified it manually like so: 

image

In the above you can see that using `inspect` directly does not work but using zenml's functon works in a notebook. Not sure how to test though.

* What do you think about the monkeypatching logic? Any other way which is more elegant?

Thanks!

### CODE REVIEW

File: src/zenml/utils/source_utils.py [hunk 1]
```diff
@@ -181,6 +190,94 @@ def get_module_source_from_class(
...
+            # If parent module is __main__, lookup by methods
+            for name, member in inspect.getmembers(object):
+                if (
+                    inspect.isfunction(member)
+                    and object.__qualname__ + "." + member.__name__
+                    == member.__qualname__
+                ):
+                    return inspect.getfile(member)
+            else:
+                raise TypeError("Source for {!r} not found".format(object))
```
Comment (schustmi) [inline src/zenml/utils/source_utils.py:255]:
```suggestion
                raise TypeError(f"Source for {object!r} not found.")
```

File: src/zenml/utils/source_utils.py [hunk 2]
```diff
@@ -181,6 +190,94 @@ def get_module_source_from_class(
...
+    return src
+
+
+def get_hashed_source(value: Any) -> str:
+    """Returns a hash of the objects source code."""
+    try:
+        source_code = get_source(value)
+    except TypeError:
+        raise TypeError(
+            "Unable to compute the hash of the source code of the step. "
```
Comment (schustmi) [inline src/zenml/utils/source_utils.py:274]: I feel like this error message should be more generic now that the method is separated from any step/pipeline code and might be used in different scenarios

**Review by schustmi — CHANGES_REQUESTED** [review]
LGTM, just two minor suggestions for error messages

**Comment by htahir1** [comment]
@schustmi Thanks, i added both suggestions. Did you perhaps have some thoughts on the questions I asked in the description of the PR? Regarding tests

**Comment by AlexejPenner** [comment]
> 

@htahir1 you could try manually setting the '__file__' attribute to none. I experimented a bit locally, this might work:


```python
import inspect
import sys


class Cat:
    a: str = 'Dog'


print("This works: \n", inspect.getsource(Cat))
setattr(sys.modules.get(Cat.__module__), '__file__', None)
print("This fails: \n", inspect.getsource(Cat))

```

**Comment by schustmi** [comment]
@htahir1 Regarding testing I think Alexej's solution seems reasonable, you'd probably have to additionally extract the `_running_in_notebook` function and mock it to return `True` during this test.

Another potentially easier solution instead of monkeypatching `inspect.getfile` could be to actually set `__file__` on the class and remove it after the call to `inspect.getsource()` finishes? Sort of the reverse approach

**Comment by htahir1** [comment]
@AlexejPenner I tried it - it indeed replicates the `inspect.getsource` error but then it makes my code fail in another way which wouldnt happen in a notebook. Therefore i left the test out. I think manual verification is all we can do for now.

@schustmi Setting the `__file__` is not really the problem I think. Its more about how inspect parses that to get the content of the source code out. I wouldnt know what to set the `__file__` to. Any suggestions?

Generally, I now used `Environment.in_notebook` method and this branch would be ready for review again  guys. Let me know if further updates are required otherwise ill merge :-)!

**Comment by schustmi** [comment]
I think in case of a class (which is the problematic case solved by this PR), inspect uses `class.__module__.__file__` in `inspect.get_file()`. If we either set `__module__` on the class or the `__file__` of this module, the default inspect implementation should also work. To get the actual file we could use the logic implemented in the PR, basically computing it from the `__file__` attribute of the methods of the class. So essentially the solution would be using similar code, but not monkeypatching the `inspect.getsource`.
However I don't think that implementation has any major upsides on the one you implemented, so no need to change it.

I do have one minor comment that just popped into my head while writing this:
I think we should make sure we always set back the original `inspect.getfile` in case someone catches an error thrown by `_new_getfile`:
```python
try:
   src = inspect.getsource(value)
finally:
   inspect.getfile = _old_getfile
```

**Comment by htahir1** [comment]
@schustmi Done! :-) Great catch!

**Review by schustmi — APPROVED** [review]

**Comment by htahir1** [comment]
@AlexejPenner Any other feedback? :-) Otherwise i can merge

**Review by AlexejPenner — APPROVED** [review]
Good to go from my side :rocket:
